homeassistant/components/nasweb/sensor.py

```diff
@@ -0,0 +1,211 @@
+    SensorDeviceClass,
+    SensorEntity,
+    SensorStateClass,
+    UnitOfTemperature,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+import homeassistant.helpers.entity_registry as er
+from homeassistant.helpers.typing import DiscoveryInfoType
+from homeassistant.helpers.update_coordinator import (
+    BaseCoordinatorEntity,
+    BaseDataUpdateCoordinatorProtocol,
+)
+
+from . import NASwebConfigEntry
+from .const import (
+    DOMAIN,
+    STATE_FAULT,
+    STATE_TAMPER,
+    STATE_UNDEFINED,
+    STATE_VIGIL,
+    STATE_VIOLATION,
+    STATUS_UPDATE_MAX_TIME_INTERVAL,
+)
+from .coordinator import NASwebCoordinator
+
+SENSOR_INPUT_TRANSLATION_KEY = "sensor_input"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _get_input(coordinator: NASwebCoordinator, index: int) -> NASwebInput | None:
+    for entry in coordinator.webio_api.inputs:
+        if entry.index == index:
+            return entry
+    return None
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config: NASwebConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up Sensor platform."""
+    coordinator = config.runtime_data
+    current_inputs: set[int] = set()
+
+    @callback
+    def _check_entities() -> None:
+        received_inputs = {entry.index for entry in coordinator.webio_api.inputs}
+        added = {i for i in received_inputs if i not in current_inputs}
+        removed = {i for i in current_inputs if i not in received_inputs}
+        entities_to_add: list[InputStateSensor] = []
+        for index in added:
+            webio_input = _get_input(coordinator, index)
+            if not isinstance(webio_input, NASwebInput):
+                _LOGGER.error("Cannot create InputStateSensor without NASwebInput")
+                continue
+            new_input = InputStateSensor(coordinator, webio_input)
+            entities_to_add.append(new_input)
+            current_inputs.add(index)
+        async_add_entities(entities_to_add)
+        entity_registry = er.async_get(hass)
+        for index in removed:
+            unique_id = f"{DOMAIN}.{config.unique_id}.input.{index}"
+            if entity_id := entity_registry.async_get_entity_id(
+                DOMAIN_SENSOR, DOMAIN, unique_id
+            ):
+                entity_registry.async_remove(entity_id)
+                current_inputs.remove(index)
+            else:
+                _LOGGER.warning("Failed to remove old input: no entity_id")
+
+    coordinator.async_add_listener(_check_entities)
+    _check_entities()
+
+    nasweb_temp_sensor = coordinator.data[KEY_TEMP_SENSOR]
+    temp_sensor = TemperatureSensor(coordinator, nasweb_temp_sensor)
+    async_add_entities([temp_sensor])
+
+
+class InputStateSensor(SensorEntity, BaseCoordinatorEntity):
```

Comment:
I think it would be useful to create a base sensor entity to deduplicate a lot of logic